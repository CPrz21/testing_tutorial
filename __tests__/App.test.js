/**
 * @format
 */

// import 'react-native';
// import React from 'react';
// import App from '../App';

// // Note: test renderer must be required after react-native.
// import renderer from 'react-test-renderer';

// it('renders correctly', () => {
//   renderer.create(<App />);
// });

import React from 'react';
import App from '../App';
import {fireEvent, render} from '@testing-library/react-native';

describe('<App/> Test', () => {
  test('should show component correctly', () => {
    const tree = render(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should create an item', () => {
    const {getByText, getByPlaceholderText} = render(<App />);
    const textInput = getByPlaceholderText('Write something');
    const addItemButton = getByText('+');
    const itemText = 'First todo';
    // Simulate adding a todo
    fireEvent.changeText(textInput, itemText);
    fireEvent.press(addItemButton);

    const createdItem = getByText(itemText);
    expect(createdItem).not.toBeNull();
  });

  test('Should create multiple items', () => {
    const {getByText, getByPlaceholderText} = render(<App />);

    const addItemButton = getByText('+');
    const textInput = getByPlaceholderText('Write something');
    const createdItemText = 'first todo';
    const createdItemText_2 = 'second todo';

    // Simulate adding a todo 1
    fireEvent.changeText(textInput, createdItemText);
    fireEvent.press(addItemButton);
    // Simulate adding a todo 2
    fireEvent.changeText(textInput, createdItemText_2);
    fireEvent.press(addItemButton);

    const firstCreatedItem = getByText(createdItemText);
    const secondCreatedItem = getByText(createdItemText_2);

    expect(firstCreatedItem).not.toBeNull();
    expect(secondCreatedItem).not.toBeNull();
  });

  test('Should delete an item', () => {
    const {getByText, getByPlaceholderText, queryByText} = render(<App />);

    const addItemButton = getByText('+');
    const textInput = getByPlaceholderText('Write something');
    const createdItemText = 'first todo';
    // Simulate adding a todo
    fireEvent.changeText(textInput, createdItemText);
    fireEvent.press(addItemButton);

    const deleteItemButton = getByText('X');
    // Simulate deleting a todo
    fireEvent.press(deleteItemButton);

    const deletedItem = queryByText(createdItemText);

    expect(deletedItem).toBeNull();
  });

  test('Should display an error when trying to create an item without any text', () => {
    const {getByText} = render(<App />);
    // Simulate adding a todo with empty input
    const addItemButton = getByText('+');
    fireEvent.press(addItemButton);

    const errorMessage = getByText('Please insert a valid text');
    expect(errorMessage).not.toBeNull();
  });

  test('Should remove the error message after creating a valid item', () => {
    const {getByText, getByPlaceholderText, queryByText} = render(<App />);

    const addItemButton = getByText('+');
    // Simulate adding a todo with empty input
    fireEvent.press(addItemButton);

    const createdItemText = 'first todo';
    const textInput = getByPlaceholderText('Write something');
    // Simulate adding a todo
    fireEvent.changeText(textInput, createdItemText);
    fireEvent.press(addItemButton);

    const errorMessage = queryByText('Please insert a valid text');
    expect(errorMessage).toBeNull();
  });
});
