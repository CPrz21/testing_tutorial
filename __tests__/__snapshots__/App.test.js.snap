// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> Test should show component correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 2,
        "justifyContent": "flex-end",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "width": "100%",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Write something"
        rejectResponderTermination={true}
        style={
          Object {
            "borderBottomColor": "black",
            "borderBottomWidth": 1,
            "width": 275,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(130, 170, 255)",
            "borderRadius": 50,
            "marginLeft": 15,
            "opacity": 1,
            "paddingHorizontal": 20,
            "paddingVertical": 3,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 32,
            }
          }
        >
          +
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 8,
        "marginVertical": 15,
        "paddingTop": 20,
        "width": "80%",
      }
    }
  >
    <View
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <RCTScrollView
        data={Array []}
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View />
      </RCTScrollView>
    </View>
    <View />
  </View>
</View>
`;
